pkg/capabilities/utils.go

```diff
@@ -0,0 +1,145 @@
+package capabilities
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+)
+
+func UnwrapRequest(request CapabilityRequest, config proto.Message, value proto.Message) (bool, error) {
+	migrated, err := FromValueOrAny(request.Inputs, request.Request, value)
+	if err != nil {
+		return migrated, err
+	}
+
+	_, err = FromValueOrAny(request.Config, request.ConfigAny, config)
+	if err != nil {
+		return migrated, err
+	}
+
+	return migrated, nil
+}
+
+func UnwrapResponse(response CapabilityResponse, value proto.Message) (bool, error) {
+	migrated, err := FromValueOrAny(response.Value, response.ResponseValue, value)
+	if err != nil {
+		return migrated, err
+	}
+
+	return migrated, nil
+}
+
+func SetResponse(response *CapabilityResponse, migrated bool, value proto.Message) error {
+	if migrated {
+		wrapped, err := anypb.New(value)
+		if err != nil {
+			return err
+		}
+		response.ResponseValue = wrapped
+		return nil
+	}
+
+	wrapped, err := values.WrapMap(value)
+	if err != nil {
+		return err
+	}
+
+	response.Value = wrapped
+	return nil
+}
+
+func FromValueOrAny(value values.Value, any *anypb.Any, into proto.Message) (bool, error) {
+	if any == nil {
+		if value == nil {
+			return false, errors.New("neither value nor any provided")
+		}
+		err := value.UnwrapTo(into)
+		return false, err
+	}
+
+	err := any.UnmarshalTo(into)
+	return true, err
+}
+
+func Execute[I, C, O proto.Message](
+	ctx context.Context,
+	request CapabilityRequest,
+	input I,
+	config C,
+	exec func(context.Context, RequestMetadata, I, C) (O, error)) (CapabilityResponse, error) {
+
+	response := CapabilityResponse{}
+	migrated, err := UnwrapRequest(request, config, input)
+	if err != nil {
+		return response, fmt.Errorf("error when unwrapping request: %w", err)
+	}
+
+	output, err := exec(ctx, request.Metadata, input, config)
+	if err != nil {
+		return response, err
+	}
+
+	if err = SetResponse(&response, migrated, output); err != nil {
+		return response, fmt.Errorf("error when setting response: %w", err)
+	}
+
+	return response, nil
+}
+
+type TriggerAndId[T proto.Message] struct {
+	Trigger T
+	Id      string
+}
+
+func RegisterTrigger[C, O proto.Message](
```

Comment:
what is `C` supposed to be? instead of say `M` (message) or `I` (input)

Reply:
Config type because of what the name was in the DAG world. Renamed to I so it's I and O like the capability request.